src/symbol/source.rs

```diff
@@ -78,3 +81,159 @@ impl super::Dwarf {
             .to_string())
     }
 }
+
+pub type CrabResult<T> = Result<T, Box<dyn std::error::Error>>;
+
+/// A line in a source code is represented as a line number and the string.
+#[derive(Debug)]
+struct SourceLine {
+    line_no: usize,
+    line_str: String,
+}
+
+#[derive(Debug)]
+/// This represents a snippet of source code. The snippet usually consists of a key line and some
+/// lines of context around it.
+pub struct Snippet {
+    /// The full path to the file, whose snippet we capture.
+    file_path: String,
+    /// The name of the function containing the breakpoint.
+    fn_name: String,
+    /// The lines from the source file that is part of the snippet. Lines are a tuple of line
+    /// number and the line string
+    lines: Vec<SourceLine>,
+    /// The index of the key line in the `lines` vector. If there is no specified key, then it will
+    /// be 0.
+    key_line_idx: usize,
+    /// The column containing the symbol we are interested in.
+    key_column_idx: usize,
+}
+
+impl Snippet {
+    /// This creates a source code snippet given the source file and the key line and the
+    /// surrounding line count as context.
+    pub fn from_file(
+        file_path: &str,
+        fn_name: String,
+        line_no: usize,
+        lines_as_context: usize,
+        column: usize,
+    ) -> CrabResult<Self> {
+        if line_no == 0 {
+            return Err("Line numbers should start at 1.".into());
+        }
+        let file = File::open(file_path)?;
+        let reader = BufReader::new(file);
+        let mut lines = vec![];
+
+        // The line number from the debuginfo starts at 1 but the one for iterator gives by
+        // `reader.lines()` starts at 0.
+        let key = line_no - 1;
+        let start = if key > lines_as_context {
+            key - lines_as_context
+        } else {
+            0
+        };
+        let end = key + lines_as_context;
+        for (i, line) in reader.lines().enumerate().skip(start) {
+            if i <= end {
+                lines.push(SourceLine {
+                    line_no: i + 1,
+                    line_str: line?,
+                });
+            } else {
+                // If we have printed the asked for line and the context around it,
+                // There is no point in iterating through the whole file. So, break
+                // out of the loop and make an early return.
+                break;
+            }
+        }
+        Ok(Snippet {
+            file_path: file_path.to_string(),
+            fn_name,
+            lines,
+            key_line_idx: lines_as_context,
+            key_column_idx: column - 1,
+        })
+    }
+}
+
+/// This module has methods to syntax highlighting of the code snippet.
+/// This is an adaptation of [this](https://github.com/bjorn3/pretty_backtrace/blob/c8a19b94f369a2edd9d58b07eb4310d2e0a2c991/src/display_frame.rs#L103-L113).
+#[cfg(feature = "syntax-highlighting")]
+pub mod pretty {
```

Comment:
```suggestion
mod pretty {
```

This can be private. Module publicness doesn't affect if methods are public.

Reply:
ahh ! I learned that now :) . WIll make the change.